draco/analysis/delay.py

```diff
@@ -734,30 +776,59 @@ def _evaluate(self, data_view, weight_view, out_cont, delays, channel_ind):
             weight = weight_view.local_array[lbi]
 
             # Apply the cuts to the data
-            t = self._cut_data(data, weight, channel_ind)
+            t = self._cut_data(data, weight)
             if t is None:
                 continue
-            data, weight, non_zero_channel = t
+            data, weight, nzf, _ = t
+
+            if self.maxpost:
+                spec, success = delay_power_spectrum_maxpost(
+                    data,
+                    ndelay,
+                    weight,
+                    None,
+                    window=self.window if self.apply_window else None,
+                    fsel=channel_ind[nzf],
+                    maxiter=self.nsamp,
+                    tol=self.maxpost_tol,
+                )
 
-            spec = delay_power_spectrum_gibbs(
-                data,
-                ndelay,
-                weight,
-                initial_S[lbi],
-                window=self.window if self.apply_window else None,
-                fsel=non_zero_channel,
-                niter=self.nsamp,
-                rng=rng,
-                complex_timedomain=self.complex_timedomain,
-            )
+                # If max-likelihood didn't converge in allowed number of iters, reflect this in the mask.
+                if not success:
+                    out_cont.datasets["spectrum_mask"][bi] = 1
```

Comment:
The indexing here is very subtle and probably worth making a comment about (I spent awhile reading it until I actually understood what is happening). When using `mpiarray.enumerate`, we get both the local array index `lbi` and the global array index `bi`. We use the local index `lbi` to get the `data`, `weight`, and `initial_S` datasets per iteration, which is expected. However, when setting the datasets in `out_cont`, we're using the global slice. This only works because we're actually indexing the `MemDataset` in `out_cont`, *not* the `MPIArray`, and `__getitem__` and `__setitem__` on `MemDatasetDistributed` operate on a global slice of the underlying array rather than the local data. So, for example, what is being done here is equivalent to doing `out_cont.datasets["spectrum_mask"][:][lbi]`, because indexing directly on the `MPIArray` is done to the local array. 

There isn't anything wrong with doing it this way, and it's consistent with what was done before. I just think there needs to be a comment here because it's quite confusing. I'm happy to add that comment later once all changes are pushed.

Reply:
Fixed in latest commit!